compiler/src/org.graalvm.compiler.core.common/src/org/graalvm/compiler/core/common/spi/JavaConstantFieldProvider.java

```diff
@@ -109,9 +113,39 @@ protected boolean isStableField(ResolvedJavaField field, ConstantFieldTool<?> to
         if (field.equals(stringHashField)) {
             return true;
         }
+        if (Options.ConsiderScalaCompanionObjectsStable.getValue(tool.getOptions()) &&
+                        isScalaCompanionObjectField(field)) {
+            return true;
+        }
+        return false;
+    }
+
+    private static final boolean isScalaCompanionObjectField(ResolvedJavaField field) {
+        ResolvedJavaType declaringClass = field.getDeclaringClass();
+        String declaringClassName = declaringClass.toClassName();
+        if (declaringClass.isInitialized() && field.getName().equals("MODULE$") && declaringClassName.endsWith("$")) {
+            try {
+                String companionInstanceClassName = declaringClassName.substring(0, declaringClassName.length() - 1);
+                Class<?> companionInstanceClass = ClassLoader.getSystemClassLoader().loadClass(companionInstanceClassName);
```

Comment:
I don't think this code is correct. You really want the class loader of `declaringClass` which is not exposed by JVMCI.

Reply:
How can I get a hold of the correct class loader? Or is it not possible? I used the system class loader because it seems that's how other parts of the code base load classes

Reply:
Class loaders are intentionally not part of JVMCI as JVMCI must also operates in environments where class loaders may not be accessible (e.g. SVM and libgraal). What you want is the API introduced recently: https://github.com/graalvm/graal-jvmci-8/blob/62cf614220f2737d775fd655b3ec0ef6715a1f04/jvmci/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotJVMCIRuntime.java#L445. Note that this API is HotSpot specific.

Reply:
@dougxc according to @retronym, we could avoid loading the class here if we read the `Scala` attribute from `declaringClass`. Is there a way to access class atributes? I couldn't find anything looking at `ResolvedJavaType`